File: mcp/ham_bands.py
"""Amateur radio band definitions and analysis helpers.

Band data is loaded from shared/ham-bands.json which is consumed by both
this Python module and the TypeScript frontend — eliminating duplication.
"""

from __future__ import annotations

import json
import math
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal, Sequence

if __package__:
    from .utils import get_field
else:
    from utils import get_field  # type: ignore[no-redef]

Region = Literal["all", "r1", "r2", "r3"]
BandQuality = Literal["excellent", "good", "marginal", "poor", "not_simulated"]


@dataclass(frozen=True, slots=True)
class HamBand:
    """A standard amateur radio band allocation."""

    label: str
    name: str
    start_mhz: float
    stop_mhz: float
    center_mhz: float
    region: Region


@dataclass(frozen=True, slots=True)
class BandPerformance:
    """Performance summary for one band."""

    band: HamBand
    simulated: bool
    point_count: int
    min_swr: float | None
    min_swr_freq_mhz: float | None
    usable_bandwidth_khz: int | None
    avg_gain_dbi: float | None
    peak_gain_dbi: float | None
    quality: BandQuality


def _find_shared_path() -> Path:
    """Locate the shared/ data directory.

    Checks two locations to support both Docker (shared/ copied alongside
    the server files) and development (shared/ at the repository root).
    """
    here = Path(__file__).parent
    # Docker layout: shared/ is copied to /opt/antsim_mcp/shared/
    docker_path = here / "shared"
    if docker_path.is_dir():
        return docker_path
    # Dev layout: shared/ is at the repo root (one level above mcp/)
    dev_path = here.parent / "shared"
    if dev_path.is_dir():
        return dev_path
    raise FileNotFoundError(
        "Could not find shared/ directory.\n"
        f"Searched:\n  {docker_path}\n  {dev_path}\n"
        "Ensure the shared/ directory exists alongside or above the mcp/ directory."
    )


def _load_ham_bands() -> list[HamBand]:
    """Load ham band definitions from shared/ham-bands.json."""
    bands_file = _find_shared_path() / "ham-bands.json"
    with open(bands_file, encoding="utf-8") as f:
        raw = json.load(f)
    return [
        HamBand(
            label=entry["label"],
            name=entry["name"],
            start_mhz=float(entry["start_mhz"]),
            stop_mhz=float(entry["stop_mhz"]),
            center_mhz=float(entry["center_mhz"]),
            region=entry["region"],
        )
        for entry in raw
    ]


# Loaded at module initialisation — replaces the hardcoded list.
HAM_BANDS: list[HamBand] = _load_ham_bands()


def _js_round(value: float, digits: int = 0) -> float:
    """Round like JavaScript's Math.round, including midpoint behaviour."""
    factor = 10 ** digits
    scaled = value * factor
    if scaled >= 0:
        return math.floor(scaled + 0.5) / factor
    return math.ceil(scaled - 0.5) / factor


def compute_steps(start_mhz: float, stop_mhz: float) -> int:
    """Compute sweep points exactly like the TypeScript computeSteps helper."""
    bw = abs(stop_mhz - start_mhz)
    return int(max(21, min(101, _js_round(bw * 25) + 1)))


def get_bands_for_region(region: Literal["r1", "r2", "r3"] = "r1") -> list[HamBand]:
    """Return bands matching the requested ITU region or 'all'."""
    if region not in {"r1", "r2", "r3"}:
        raise ValueError("region must be one of: r1, r2, r3")
    return [band for band in HAM_BANDS if band.region == "all" or band.region == region]


def get_band_by_label(label: str, region: Literal["r1", "r2", "r3"] = "r1") -> HamBand:
    """Look up a band by short label or full name.

    Region-specific matches are preferred for the requested region.
    """
    normalized = label.strip().casefold()
    for band in get_bands_for_region(region):
        if band.label.casefold() == normalized or band.name.casefold() == normalized:
            return band
    for band in HAM_BANDS:
        if band.label.casefold() == normalized or band.name.casefold() == normalized:
            return band
    raise ValueError(f"Unknown ham band: {label!r}")


def band_to_frequency_range(band: HamBand) -> dict[str, float | int]:
    """Convert a band definition to a default sweep range."""
    return {
        "start_mhz": band.start_mhz,
        "stop_mhz": band.stop_mhz,
        "steps": compute_steps(band.start_mhz, band.stop_mhz),
    }


def analyze_band_performance(
    results: Sequence[Any],
    region: Literal["r1", "r2", "r3"] = "r1",
    swr_threshold: float = 2.0,
) -> list[BandPerformance]:
    """Analyse simulation results across all defined bands for one region.

    Mirrors the TypeScript analyzeBandPerformance logic:
    - find all frequency points that fall inside each band
    - determine min SWR and where it occurs
    - estimate usable bandwidth where SWR <= threshold
    - compute average and peak gain across the band
    - assign a quality rating based on minimum SWR
    """
    performances: list[BandPerformance] = []

    for band in get_bands_for_region(region):
        in_band = [
            result
            for result in results
            if band.start_mhz <= float(get_field(result, "frequency_mhz")) <= band.stop_mhz
        ]

        if not in_band:
            performances.append(
                BandPerformance(
                    band=band,
                    simulated=False,
                    point_count=0,
                    min_swr=None,
                    min_swr_freq_mhz=None,
                    usable_bandwidth_khz=None,
                    avg_gain_dbi=None,
                    peak_gain_dbi=None,
                    quality="not_simulated",
                )
            )
            continue

        min_swr = math.inf
        min_swr_freq = 0.0
        for result in in_band:
            swr = float(get_field(result, "swr_50"))
            if swr < min_swr:
                min_swr = swr
                min_swr_freq = float(get_field(result, "frequency_mhz"))

        usable = [
            result
            for result in in_band
            if float(get_field(result, "swr_50")) <= swr_threshold
        ]
        usable_bandwidth_khz: int | None = None
        if usable:
            min_freq = min(float(get_field(r, "frequency_mhz")) for r in usable)
            max_freq = max(float(get_field(r, "frequency_mhz")) for r in usable)
            usable_bandwidth_khz = int(_js_round((max_freq - min_freq) * 1000))

        gains = [
            float(get_field(result, "gain_max_dbi"))
            for result in in_band
            if float(get_field(result, "gain_max_dbi")) > -999.0
        ]
        avg_gain = sum(gains) / len(gains) if gains else None
        peak_gain = max(gains) if gains else None

        if min_swr <= 1.5:
            quality: BandQuality = "excellent"
        elif min_swr <= 2.0:
            quality = "good"
        elif min_swr <= 3.0:
            quality = "marginal"
        else:
            quality = "poor"

        performances.append(
            BandPerformance(
                band=band,
                simulated=True,
                point_count=len(in_band),
                min_swr=_js_round(min_swr, 2),
                min_swr_freq_mhz=_js_round(min_swr_freq, 3),
                usable_bandwidth_khz=usable_bandwidth_khz,
                avg_gain_dbi=_js_round(avg_gain, 2) if avg_gain is not None else None,
                peak_gain_dbi=_js_round(peak_gain, 2) if peak_gain is not None else None,
                quality=quality,
            )
        )

    return performances


__all__ = [
    "BandPerformance",
    "HAM_BANDS",
    "HamBand",
    "analyze_band_performance",
    "band_to_frequency_range",
    "compute_steps",
    "get_band_by_label",
    "get_bands_for_region",
]

File: frontend/src/utils/ham-bands.ts
/**
 * Amateur radio band definitions and analysis utilities.
 *
 * Band data is loaded from shared/ham-bands.json — the single source of truth
 * consumed by both this TypeScript module and the Python MCP server.
 */

import type { FrequencyRange, FrequencySegment } from "../templates/types";
import type { FrequencyResult } from "../api/nec";
import rawBandsData from "../../../shared/ham-bands.json";

// ---------------------------------------------------------------------------
// Band definitions
// ---------------------------------------------------------------------------

export interface HamBand {
  /** Short label: "40m", "20m", etc. */
  label: string;
  /** Full name: "40 meters" */
  name: string;
  /** Lower band edge in MHz */
  start_mhz: number;
  /** Upper band edge in MHz */
  stop_mhz: number;
  /** Center frequency in MHz */
  center_mhz: number;
  /** ITU region: "all" = worldwide, "r1" = Europe/Africa, "r2" = Americas, "r3" = Asia/Pacific */
  region: "all" | "r1" | "r2" | "r3";
}

// Type of the raw JSON entries (region is widened to string by JSON import)
interface RawHamBand {
  label: string;
  name: string;
  start_mhz: number;
  stop_mhz: number;
  center_mhz: number;
  region: string;
}

/**
 * Standard amateur radio HF/VHF/UHF bands.
 *
 * Loaded from shared/ham-bands.json — both Python (MCP server) and TypeScript
 * consume the same file, so changes need to be made in only one place.
 *
 * For bands where Region 1/2/3 allocations differ (80m, 40m), the JSON
 * contains separate entries for each region.
 */
export const HAM_BANDS: HamBand[] = (rawBandsData as RawHamBand[]).map((b) => ({
  label: b.label,
  name: b.name,
  start_mhz: b.start_mhz,
  stop_mhz: b.stop_mhz,
  center_mhz: b.center_mhz,
  region: b.region as HamBand["region"],
}));

/**
 * Get bands for a specific ITU region.
 * Returns bands where region matches or is "all".
 * For bands with region-specific variants (80m, 40m), only the matching variant is returned.
 */
export function getBandsForRegion(region: "r1" | "r2" | "r3" = "r1"): HamBand[] {
  return HAM_BANDS.filter((b) => b.region === "all" || b.region === region);
}

/**
 * Get the band-edge-only list used by SWRChart for marking band edges.
 * Returns one entry per unique label (no duplicates from region variants).
 */
export function getBandEdges(region: "r1" | "r2" | "r3" = "r1"): Array<{ name: string; start: number; end: number }> {
  return getBandsForRegion(region).map((b) => ({
    name: b.label,
    start: b.start_mhz,
    end: b.stop_mhz,
  }));
}

// ---------------------------------------------------------------------------
// Frequency range from band
// ---------------------------------------------------------------------------

/**
 * Compute a sensible number of sweep steps for a given frequency range.
 *
 * Uses ~25 points per MHz of bandwidth, clamped to [21, 101].
 */
export function computeSteps(startMhz: number, stopMhz: number): number {
  const bw = Math.abs(stopMhz - startMhz);
  return Math.max(21, Math.min(101, Math.round(bw * 25) + 1));
}

/**
 * Create a FrequencyRange from a ham band.
 * Steps are computed automatically from the bandwidth.
 */
export function bandToFrequencyRange(band: HamBand): FrequencyRange {
  return {
    start_mhz: band.start_mhz,
    stop_mhz: band.stop_mhz,
    steps: computeSteps(band.start_mhz, band.stop_mhz),
  };
}

// ---------------------------------------------------------------------------
// Multi-segment helpers
// ---------------------------------------------------------------------------

/** Convert a HamBand to a FrequencySegment */
export function bandToSegment(band: HamBand): FrequencySegment {
  return {
    start_mhz: band.start_mhz,
    stop_mhz: band.stop_mhz,
    steps: computeSteps(band.start_mhz, band.stop_mhz),
    label: band.label,
  };
}

/** Check if a segment matches a band's frequency range */
function segmentMatchesBand(seg: FrequencySegment, band: HamBand): boolean {
  return (
    Math.abs(seg.start_mhz - band.start_mhz) < 0.01 &&
    Math.abs(seg.stop_mhz - band.stop_mhz) < 0.01
  );
}

/** Check if a band already exists in the segments list */
export function hasBandSegment(segments: FrequencySegment[], band: HamBand): boolean {
  return segments.some((seg) => segmentMatchesBand(seg, band));
}

/** Remove a band's segment from the list */
export function removeBandSegment(segments: FrequencySegment[], band: HamBand): FrequencySegment[] {
  return segments.filter((seg) => !segmentMatchesBand(seg, band));
}

// ---------------------------------------------------------------------------
// Band performance analysis
// ---------------------------------------------------------------------------

export interface BandPerformance {
  band: HamBand;
  simulated: boolean;
  pointCount: number;
  minSwr: number | null;
  minSwrFreqMhz: number | null;
  usableBandwidthKhz: number | null;
  avgGainDbi: number | null;
  peakGainDbi: number | null;
  quality: "excellent" | "good" | "marginal" | "poor" | "not_simulated";
}

/**
 * Analyze simulation results across all ham bands for a given region.
 */
export function analyzeBandPerformance(
  results: FrequencyResult[],
  region: "r1" | "r2" | "r3" = "r1",
  swrThreshold: number = 2.0,
): BandPerformance[] {
  const bands = getBandsForRegion(region);

  return bands.map((band) => {
    const inBand = results.filter(
      (r) => r.frequency_mhz >= band.start_mhz && r.frequency_mhz <= band.stop_mhz,
    );

    if (inBand.length === 0) {
      return {
        band,
        simulated: false,
        pointCount: 0,
        minSwr: null,
        minSwrFreqMhz: null,
        usableBandwidthKhz: null,
        avgGainDbi: null,
        peakGainDbi: null,
        quality: "not_simulated" as const,
      };
    }

    let minSwr = Infinity;
    let minSwrFreq = 0;
    for (const r of inBand) {
      if (r.swr_50 < minSwr) {
        minSwr = r.swr_50;
        minSwrFreq = r.frequency_mhz;
      }
    }

    const usable = inBand.filter((r) => r.swr_50 <= swrThreshold);
    let usableBwKhz: number | null = null;
    if (usable.length > 0) {
      const minFreq = Math.min(...usable.map((r) => r.frequency_mhz));
      const maxFreq = Math.max(...usable.map((r) => r.frequency_mhz));
      usableBwKhz = Math.round((maxFreq - minFreq) * 1000);
    }

    const gains = inBand.map((r) => r.gain_max_dbi).filter((g) => g > -999);
    const avgGain = gains.length > 0 ? gains.reduce((a, b) => a + b, 0) / gains.length : null;
    const peakGain = gains.length > 0 ? Math.max(...gains) : null;

    let quality: BandPerformance["quality"];
    if (minSwr <= 1.5) {
      quality = "excellent";
    } else if (minSwr <= 2.0) {
      quality = "good";
    } else if (minSwr <= 3.0) {
      quality = "marginal";
    } else {
      quality = "poor";
    }

    return {
      band,
      simulated: true,
      pointCount: inBand.length,
      minSwr: Math.round(minSwr * 100) / 100,
      minSwrFreqMhz: Math.round(minSwrFreq * 1000) / 1000,
      usableBandwidthKhz: usableBwKhz,
      avgGainDbi: avgGain !== null ? Math.round(avgGain * 100) / 100 : null,
      peakGainDbi: peakGain !== null ? Math.round(peakGain * 100) / 100 : null,
      quality,
    };
  });
}